cmd/bucket-replication.go

```diff
@@ -2675,3 +2688,82 @@ func (p *ReplicationPool) queueMRFHeal(file string) error {
 	}
 	return nil
 }
+
+// load replication stats from disk
+func (p *ReplicationPool) loadStatsFromDisk() (rs map[string]BucketReplicationStats, e error) {
+	data, err := readConfig(p.ctx, p.objLayer, replicationStatsPath)
+	if err != nil && err != errConfigNotFound {
+		return rs, err
+	}
+
+	if len(data) <= 4 {
+		return map[string]BucketReplicationStats{}, nil
+	}
+	// Read repl stats meta header
+	switch binary.LittleEndian.Uint16(data[0:2]) {
+	case replStatsMetaFormat:
+	default:
+		return rs, fmt.Errorf("replication stats: unknown format: %d", binary.LittleEndian.Uint16(data[0:2]))
+	}
+	switch binary.LittleEndian.Uint16(data[2:4]) {
+	case replStatsVersion:
+	default:
+		return rs, fmt.Errorf("replication stats: unknown version: %d", binary.LittleEndian.Uint16(data[2:4]))
+	}
+	ss := BucketStatsMap{}
+	if _, err = ss.UnmarshalMsg(data[4:]); err != nil {
+		return rs, err
+	}
+	rs = make(map[string]BucketReplicationStats, len(ss))
+	for bucket, st := range ss {
+		rs[bucket] = st.ReplicationStats
+	}
+
+	return rs, nil
+}
+
+func (p *ReplicationPool) saveStatsToDisk() {
+	if p == nil || p.objLayer == nil {
+		return
+	}
+	sTimer := time.NewTimer(replStatsSaveInterval)
+	defer sTimer.Stop()
+	for {
+		select {
+		case <-sTimer.C:
+			dui, err := loadDataUsageFromBackend(GlobalContext, newObjectLayerFn())
+			if err == nil && !dui.LastUpdate.IsZero() {
+				globalReplicationStats.saveUsage(dui.BucketsUsage)
+			}
+			p.saveStats(p.ctx, dui.BucketsUsage)
+			sTimer.Reset(replStatsSaveInterval)
+		case <-p.ctx.Done():
+			p.saveStats(context.Background(), globalReplicationStats.getUsage())
+			return
+		}
+	}
+}
+
+// save replication stats to .minio.sys/buckets/replication/stats.bin
+func (p *ReplicationPool) saveStats(ctx context.Context, ui map[string]BucketUsageInfo) error {
+	stats := getAllLatestReplicationStats(ui)
+	if len(stats) == 0 {
+		return nil
+	}
+	bucketStatsMap := make(map[string]BucketStats, len(stats))
+	for k, v := range stats {
+		bucketStatsMap[k] = BucketStats{
+			ReplicationStats: v,
+		}
+	}
+
+	var buf bytes.Buffer
+	msgp.Encode(&buf, BucketStatsMap(bucketStatsMap))
+	data := make([]byte, buf.Len()+4)
+
+	// Initialize the replication stats meta header.
+	binary.LittleEndian.PutUint16(data[0:2], replStatsMetaFormat)
+	binary.LittleEndian.PutUint16(data[2:4], replStatsVersion)
+	copy(data[4:], buf.Bytes())
+	return saveConfig(ctx, p.objLayer, replicationStatsPath, data)
```

Comment:
A bit simpler, with less allocs:
```suggestion
	bucketStatsMap := make(BucketStatsMap, len(stats))
	for k, v := range stats {
		bucketStatsMap[k] = BucketStats{
			ReplicationStats: v,
		}
	}

	data := make([]byte, 4, 4+bucketStatsMap.Msgsize())
	// Add the replication stats meta header.
	binary.LittleEndian.PutUint16(data[0:2], replStatsMetaFormat)
	binary.LittleEndian.PutUint16(data[2:4], replStatsVersion)
	// Add data
	data, err := bucketStatsMap.MarshalMsg(data)
	if err != nil {
		return err
	}
	return saveConfig(ctx, p.objLayer, replicationStatsPath, data)
```